File: attendai-ai-service/app/services/risk_scorer.py
"""Risk scoring.

This service exists so the AI side can offer either:
  1. The same deterministic rule-based scorer the Java backend uses (default),
     letting both sides agree on borderline cases.
  2. A swap-in ML scorer trained on flagged attendance attempts (Phase 5+).

We start with #1 - identical to RiskScorer.java on the Spring Boot side - and
return a `model` field so callers can tell which version produced the score.
"""

from __future__ import annotations

from typing import List

from ..schemas.risk import RiskLevel, RiskScoreRequest, RiskScoreResponse


# Thresholds match RiskScorer.java -> AppProperties.RiskThresholds.
_LOW = 30
_MEDIUM = 60
_HIGH = 80


def _level_for(score: int) -> RiskLevel:
    if score <= _LOW:
        return RiskLevel.LOW
    if score <= _MEDIUM:
        return RiskLevel.MEDIUM
    if score <= _HIGH:
        return RiskLevel.HIGH
    return RiskLevel.CRITICAL


def score(req: RiskScoreRequest) -> RiskScoreResponse:
    """Compute a 0..100 risk score from per-attempt signals."""
    s = 0
    factors: List[str] = []

    if req.wrong_code_attempts > 0:
        add = min(30, req.wrong_code_attempts * 10)
        s += add
        factors.append(f"wrong_code_attempts:{req.wrong_code_attempts}")

    if not req.device_trusted:
        s += 20
        factors.append("untrusted_device")

    if req.device_used_by_multiple_accounts:
        s += 30
        factors.append("multi_account_device")

    if req.face_failed:
        s += 40
        factors.append("face_failed")
    elif req.face_confidence < 0.75:
        slope = (0.75 - req.face_confidence) / 0.25
        add = int(min(25.0, slope * 25.0))
        if add > 0:
            s += add
            factors.append(f"low_face_confidence:{req.face_confidence:.2f}")

    if req.late_by_seconds > 0:
        s += 50
        factors.append(f"late_by_seconds:{req.late_by_seconds}")

    if req.duplicate_attempt:
        s += 50
        factors.append("duplicate_attempt")

    if req.missed_random_challenges > 0:
        s += min(20, req.missed_random_challenges * 5)
        factors.append(f"missed_random_challenges:{req.missed_random_challenges}")

    if req.suspicious_ip_pattern:
        s += 20
        factors.append("suspicious_ip")

    s = min(100, s)
    return RiskScoreResponse(score=s, level=_level_for(s), factors=factors)

File: attendai-backend/src/main/java/com/attendai/attendance/verification/RiskScorer.java
package com.attendai.attendance.verification;

import com.attendai.config.AppProperties;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

/**
 * Computes a 0-100 risk score for an attendance attempt. The default
 * implementation is a deterministic rule-based scorer that matches the table
 * in SRS § 14. Phase 5 will plug in a model-based variant.
 *
 * Inputs come from the validation pipeline (see {@link RiskInput}).
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class RiskScorer {

    private final AppProperties props;

    public enum Level { LOW, MEDIUM, HIGH, CRITICAL }

    public record RiskInput(
            int wrongCodeAttempts,
            boolean deviceTrusted,
            boolean deviceUsedByMultipleAccounts,
            boolean faceFailed,
            double faceConfidence,
            int lateBySeconds,
            boolean duplicateAttempt,
            int missedRandomChallenges,
            boolean suspiciousIpPattern
    ) {}

    public record RiskOutput(int score, Level level, List<String> factors) {}

    public RiskOutput score(RiskInput in) {
        int score = 0;
        List<String> factors = new ArrayList<>();

        if (in.wrongCodeAttempts() > 0) {
            int add = Math.min(30, in.wrongCodeAttempts() * 10);
            score += add;
            factors.add("wrong_code_attempts:" + in.wrongCodeAttempts());
        }
        if (!in.deviceTrusted()) {
            score += 20;
            factors.add("untrusted_device");
        }
        if (in.deviceUsedByMultipleAccounts()) {
            score += 30;
            factors.add("multi_account_device");
        }
        if (in.faceFailed()) {
            score += 40;
            factors.add("face_failed");
        } else if (in.faceConfidence() < 0.75) {
            // Linear 25-point penalty as confidence falls from 0.75 to 0.50.
            double slope = (0.75 - in.faceConfidence()) / 0.25;
            int add = (int) Math.min(25.0, slope * 25.0);
            score += Math.max(0, add);
            factors.add("low_face_confidence:" + String.format("%.2f", in.faceConfidence()));
        }
        if (in.lateBySeconds() > 0) {
            score += 50;
            factors.add("late_by_seconds:" + in.lateBySeconds());
        }
        if (in.duplicateAttempt()) {
            score += 50;
            factors.add("duplicate_attempt");
        }
        if (in.missedRandomChallenges() > 0) {
            score += Math.min(20, in.missedRandomChallenges() * 5);
            factors.add("missed_random_challenges:" + in.missedRandomChallenges());
        }
        if (in.suspiciousIpPattern()) {
            score += 20;
            factors.add("suspicious_ip");
        }

        score = Math.min(100, score);

        var thresholds = props.attendance().riskThresholds();
        Level level;
        if (score <= thresholds.low()) level = Level.LOW;
        else if (score <= thresholds.medium()) level = Level.MEDIUM;
        else if (score <= thresholds.high()) level = Level.HIGH;
        else level = Level.CRITICAL;

        log.debug("Risk score {} ({}) factors={}", score, level, factors);
        return new RiskOutput(score, level, List.copyOf(factors));
    }
}
